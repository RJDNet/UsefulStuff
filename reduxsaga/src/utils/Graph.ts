import PriorityQueue from './PriorityQueue';

export default class Graph {
  public nodes: Array<string>;
  private adjacencyList: { [index: string]: Array<Object> };

  constructor() {
    this.nodes = [];
    this.adjacencyList = {};
  }

  addNode(node: string): void {
    this.nodes.push(node);
    this.adjacencyList[node] = [];
  }

  addEdge(node1: string, node2: string, distance: number): void {
    this.adjacencyList[node1].push({ node: node2, distance: distance });
    this.adjacencyList[node2].push({ node: node1, distance: distance });
  }

  findPath(startNode: string, endNode: string) {
    let times: Record<string, number> = {};
    let backtrace: Record<string, any> = {};
    let pq = new PriorityQueue();

    times[startNode] = 0;

    this.nodes.forEach(node => {
      if (node !== startNode) {
        times[node] = Infinity
      }
    });

    pq.enqueue([startNode, 0]);

    while (!pq.isEmpty()) {
      let shortestStep = pq.dequeue();
      if (shortestStep) {
        let currentNode = shortestStep[0];
        this.adjacencyList[currentNode].forEach((neighbor: any) => {
          let time = times[currentNode] + neighbor.distance;

          if (time < times[neighbor.node]) {
            times[neighbor.node] = time;
            backtrace[neighbor.node] = currentNode;
            pq.enqueue([neighbor.node, time]);
          }
        });
      }
    }

    let path = [endNode];
    let lastStep = endNode;

    while (lastStep !== startNode) {
      path.unshift(backtrace[lastStep])
      lastStep = backtrace[lastStep]
    }

    return {
      path,
      times: times[endNode]
    };
  }

}
