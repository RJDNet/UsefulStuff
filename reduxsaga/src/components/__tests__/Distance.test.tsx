import React from 'react';
import ReactDOM from 'react-dom';

// Jest
import 'jest-dom/extend-expect';

// Enzyme
import { shallow, mount } from 'enzyme';

// React Testing Library
import { render } from 'react-testing-library';

// React Component
import Distance from '../Distance';

// Utility Component
import Graph from '../../utils/Graph';

describe('distance Component', () => {
  const map = new Graph();
  const wrapper = mount(<Distance map={map} />);

  describe('rendering component', () => {
    it('renders component without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<Distance map={map} />, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    // Snapshot can be useful for tracking changes in UI
    it('snapshot', () => {
      const wrapper = shallow(<Distance map={map} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('contains prop', () => {
    it('accepts map prop', () => {
      expect(wrapper.props().map).toEqual(map)
    });
  });

  describe('contains state', () => {
    it('contains start state key', () => {
      const wrapper = shallow(<Distance map={map} />);
      expect(wrapper.state('start')).toContain('');
    });

    it('contains start end key', () => {
      const wrapper = shallow(<Distance map={map} />);
      expect(wrapper.state('end')).toContain('');
    });

    it('contains start route key', () => {
      const wrapper = shallow(<Distance map={map} />);
      expect(wrapper.state('route')).toContain('');
    });
  });

  describe('rendering elements', () => {
    it('renders start input', async () => {
      const { getByPlaceholderText } = render(<Distance map={map} />)
      expect(getByPlaceholderText('Route Start')).toBeTruthy();
    });

    it('renders end input', async () => {
      const { getByPlaceholderText } = render(<Distance map={map} />)
      expect(getByPlaceholderText('Route End')).toBeTruthy();
    });

    it('renders from h4 element', async () => {
      const { getByTestId } = render(<Distance map={map} />)
      expect(getByTestId('from')).toBeTruthy();
    });

    it('renders end h4 element', async () => {
      const { getByTestId } = render(<Distance map={map} />)
      expect(getByTestId('end')).toBeTruthy();
    });

    it('renders shortestRoute h4 element', async () => {
      const { getByTestId } = render(<Distance map={map} />)
      expect(getByTestId('shortestRoute')).toBeTruthy();
    });

    it('renders button', async () => {
      const { getByPlaceholderText } = render(<Distance map={map} />)
      expect(getByPlaceholderText('Route End')).toBeTruthy();
    });
  });

  describe('input value change', () => {
    it('start input value should change on keypress', () => {
      const component = mount(<Distance map={map} />);
      component.find('.startRouteInput').simulate('keydown', { keyCode: 66 });
      expect(component).toMatchSnapshot();
      component.unmount();
    });

    it('end input value should change on keypress', () => {
      const component = mount(<Distance map={map} />);
      component.find('.endRouteInput').simulate('keydown', { keyCode: 72 });
      expect(component).toMatchSnapshot();
      component.unmount();
    });
  });

  describe('h4 text changes on state change', () => {
    it('h4 from element', () => {
      const wrapper = mount(<Distance map={map} />);
      expect(wrapper.find('.from').text()).toEqual('From: ');
      wrapper.setState({ start: 'B' });
      expect(wrapper.find('.from').text()).toEqual('From: B');
      wrapper.unmount();
    });

    it('h4 end element', () => {
      const wrapper = mount(<Distance map={map} />);
      expect(wrapper.find('.end').text()).toEqual('To: ');
      wrapper.setState({ end: 'H' });
      expect(wrapper.find('.end').text()).toEqual('To: H');
      wrapper.unmount();
    });

    it('h4 route element', () => {
      const wrapper = mount(<Distance map={map} />);
      expect(wrapper.find('.shortestRoute').text()).toEqual('Shortest Route: ');
      wrapper.setState({ route: 'BDGH with a distance of 10' });
      expect(wrapper.find('.shortestRoute').text()).toEqual('Shortest Route: BDGH with a distance of 10');
      wrapper.unmount();
    });
  });
});
