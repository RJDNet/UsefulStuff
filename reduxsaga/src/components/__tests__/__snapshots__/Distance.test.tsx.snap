// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`distance Component input value change end input value should change on keypress 1`] = `ReactWrapper {}`;

exports[`distance Component input value change start input value should change on keypress 1`] = `ReactWrapper {}`;

exports[`distance Component rendering component snapshot 1`] = `ShallowWrapper {}`;
